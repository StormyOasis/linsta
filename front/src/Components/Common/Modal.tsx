import React from "react";
import { createPortal } from "react-dom";
import { enableModal } from "../../utils/utils";
import styled from "styled-components";

export type ModalProps = {
  children?: any;
  title?: string;
  onClose: any;
};

const ModalWrapper = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
  justify-content: center;
  overflow: hidden;
  position: relative;
  pointer-events: all;
  z-index: 99999;
  border: none;
  box-sizing: content-box;
`;

const ModalInnerWrapper = styled.div`
  animation-duration: 0.15s;
  animation-iteration-count: 1;
  animation-timing-function: ease-out;
  animation-name: modalAnimation;
  display: block;
  flex-shrink: 1;
  margin: 20px;
  max-width: ${props => props.theme['sizes'].maxModalWidth};
  max-height: calc(100% - 40px);
  min-width: 260px;
  pointer-events: all;
  position: relative;
  width: ${props => props.theme['sizes'].defaultModalWidth};
  border-radius: 12px;
  background-color: white;
`;

const ModalInnerWrapper2 = styled.div`  
  display: block;
  overflow: auto;
  pointer-events: all;
`;

const ModalTitleBarWrapper = styled.div`
  align-items: stretch;
  display: flex;
  flex-direction: column;
  pointer-events: all;
`;

const ModalTitleBarInnerWrapper = styled.div`
  align-items: center;
  display: block;
  border-bottom: 1px solid rgb(220, 220, 220);
  height: 42px;
  pointer-events: all;
  position: relative;
  width: ${props => props.theme['sizes'].defaultModalWidth};
  max-width: width: ${props => props.theme['sizes'].maxModalWidth};
`;

const ModalTitleBarInnerWrapper2 = styled.div`
  align-content: stretch;
  align-items: center;
  box-sizing: border-box;
  display: flex;
  flex-direction: column;
  flex-grow: 0;
  flex-shrink: 0;
  height: 42px;
  justify-content: center;
  overflow: visible;
  pointer-events: all;
  width: ${props => props.theme['sizes'].defaultModalWidth};
  max-width: width: ${props => props.theme['sizes'].maxModalWidth};
  position: absolute;
`;

const ModalTitle = styled.div`
  align-items: center;
  display: flex;
  flex-direction: column;
  flex-grow: 1
  font-size: 16px;
  font-weight: 600;
  line-height: 20px;
  text-align: center;
  justify-content: center; 
  width: calc(100%-100); 
`;

const ModalCloseWrapper = styled.div`
  align-items: center;
  display: flex;
  flex-basis: 48px;
  flex-direction: column;
  float: right;
  height: 42px;
  justify-content: center;
  pointer-events: all;
  position: relative;
`;

const ModalClose = styled.div`
  align-content: stretch;
  align-items: stretch;
  border: none;
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  overflow: visible;
  padding: 0 8px;
  pointer-events: all;
  position: relative;
`;

const ModalCloseButton = styled.button`
  align-items: center;
  background-color: white;
  border: none;
  cursor: pointer;
  display: flex;
  justify-content: center;
  pointer-events: all;
  text-align: center;
`;

const CloseButton = styled.span`
  width: 12px;
  height: 12px;
  margin-right: 5px;
  background: url('/public/images/x.svg');
`;

export const ModalContentWrapper = styled.div`
    align-content: stretch;
    align-items: center;
    border: none;
    display: flex;
    flex-direction: column;
    flex-grow: 0;
    flex-shrink: 0;
    justify-content: flex-start;
    overflow: visible;
    pointer-events: all;
    position: relative;
    margin: 20px 28px 20px 28px;
`;

export const ModalSectionWrapper = styled.div`
    align-content: stretch;
    align-items: stretch;
    display: flex;
    flex-direction: column;
    flex-grow: 0;
    flex-shrink: 0;
    justify-content: flex-start;
    overflow: visible;
    position: relative;
    pointer-events: all;
`;

export default class Modal extends React.Component<ModalProps> {
  onClose = (event: React.ChangeEvent<HTMLSelectElement>) => {
    this.props.onClose && this.props.onClose(event);
    enableModal(false);
  };

  override componentDidMount(): void {
    enableModal(true);
  }

  override render() {
    const cont = document.getElementById("modalContainer");
    if(cont == null) {
      throw new Error("No modal container");
    }

    return createPortal(
      <>
        <ModalWrapper role="dialog">
          <ModalInnerWrapper>
            <ModalInnerWrapper2>
              <div style={{ display: "flex", flexDirection: "column", height: "100%", maxWidth: "100%" }}>
                <ModalTitleBarWrapper>
                  <ModalTitleBarInnerWrapper>
                    <ModalTitleBarInnerWrapper2>
                      <ModalTitle>
                        <div>{this.props.title}</div>
                      </ModalTitle>
                    </ModalTitleBarInnerWrapper2>
                    <ModalCloseWrapper>
                      <ModalClose>
                        <ModalCloseButton title="data-modal-close" aria-label="Close" onClick={this.onClose}>
                          <div style={{ alignItems: "center", display: "flex", justifyContent: "center", cursor: "pointer" }}>
                            <CloseButton />
                          </div>
                        </ModalCloseButton>
                      </ModalClose>
                    </ModalCloseWrapper>
                  </ModalTitleBarInnerWrapper>
                </ModalTitleBarWrapper>
                <ModalContentWrapper>
                  {this.props.children}
                </ModalContentWrapper>
              </div>
            </ModalInnerWrapper2>
          </ModalInnerWrapper>
        </ModalWrapper>
      </>
    , cont);
  }
}