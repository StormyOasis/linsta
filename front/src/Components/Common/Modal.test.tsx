import React from 'react'
import renderer from 'react-test-renderer';
import {render, fireEvent, screen} from '@testing-library/react';
import userEvent from '@testing-library/user-event'
import '@testing-library/jest-dom';
import Jest from 'jest';
import Modal from './Modal';


test("Modal Component", () => {
    const r = render(<Modal title="Test Modal" onClose={()=> true}>Hello</Modal>).asFragment();
    expect(r).toMatchSnapshot();
})

test("Modal Close", async () => {
    let cb = jest.fn();
    const r = render(<Modal title="Test Modal" onClose={cb}>Hello</Modal>).asFragment();
  
    await userEvent.click(screen.getByTitle('data-modal-close'));
    expect(cb).toHaveBeenCalled();
})