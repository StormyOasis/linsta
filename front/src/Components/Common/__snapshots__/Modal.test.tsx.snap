// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component 1`] = `
<DocumentFragment>
  <div
    class="Modal__ModalWrapper-sc-ehox15-0 ithfnX"
    role="dialog"
  >
    <div
      class="Modal__ModalInnerWrapper-sc-ehox15-1 hSKvcq"
    >
      <div
        class="Modal__ModalInnerWrapper2-sc-ehox15-2 Qehko"
      >
        <div
          style="display: flex; flex-direction: column; height: 100%; max-width: 100%;"
        >
          <div
            class="Modal__ModalTitleBarWrapper-sc-ehox15-3 SSVjs"
          >
            <div
              class="Modal__ModalTitleBarInnerWrapper-sc-ehox15-4 kleBih"
            >
              <div
                class="Modal__ModalTitleBarInnerWrapper2-sc-ehox15-5 ffpHMf"
              >
                <div
                  class="Modal__ModalTitle-sc-ehox15-6 flwYZd"
                >
                  <div>
                    Test Modal
                  </div>
                </div>
              </div>
              <div
                class="Modal__ModalCloseWrapper-sc-ehox15-7 funGuz"
              >
                <div
                  class="Modal__ModalClose-sc-ehox15-8 LwzWh"
                >
                  <button
                    aria-label="Close"
                    class="Modal__ModalCloseButton-sc-ehox15-9 bcIJQs"
                    title="data-modal-close"
                  >
                    <div
                      style="align-items: center; display: flex; justify-content: center; cursor: pointer;"
                    >
                      <span
                        class="Modal__CloseButton-sc-ehox15-11 feBrNy"
                      />
                    </div>
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="Modal__ModalContentWrapper-sc-ehox15-10"
          >
            Hello
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
